### ISSUE TITLE
[Request] New ItemComponent Types Definitions Assistance

### ISSUE THREAD

**Issue by MapleWheels** [Regalis11/Barotrauma#10157]
Good day,

I am aware that this request falls outside the purview of the Dev team and that any issues that arise from what I'm trying to do have no bearing of responsibility on Undertow, Fakefish, and others nor do you have any obligations to help me.

So, using the following mod "CsForBarotrauma", I am attempting to add a new child Type of Reactor in order to implement a new type of power generation (I need to override some of your methods). What I would like help with is understanding your Content Loading system. I have looked through the source and have a background in .NET so I am aware of how you're using XML (System.Xml) for this. However, I have been told by multiple people that Types are "hardcoded" so adding any new ItemComponent Subtypes is highly not advisable. However, I cannot for the life of me see where and how this is. It seems that so long as implement the interfaces and attributes as you have, simply creating the Subtype will make this work.

People have recommended using runtime editing, like Harmony, but I find that can get messy, especially if multiple people are trying to edit this Type. I think a cleaner approach would be to define my own Reactor Subtype and implement the changes I want for my mod there, keeping my changes contained and clear.

What would be your recommended way to create a new ItemComponent Type to reduce potential conflicts and if you have the time, would you mind explaining/detailing from the code side, your XML content loading system?

Thanks for any assistance you can provide.

Sincerely,

**Comment by Regalis11** [comment]
> Edit: I guess the best way to phrase this question would be "When you guys add a new Subtype to the game for use in XML content (ie. <PowerContainer>, <Powered>, etc.), what do you guys have to do?" As far as I can see, reflection is doing the work.

That's correct: all you need to do is add a new subclass of ItemComponent and the game should be able to find it using reflection.

> UPDATE: So from what I understand, LuaCsForBarotrauma does not execute any client-side code before you enter a lobby (SubEditor, CampaignLobby, MultiplayerLobby). So I would need to force a reload of all XML data from Content Packages before you join the lobby or you would desync. From what I was told, the method for this was removed in the Modding Refactor Update. Is there any new equivalent for this?

I'm not familiar with the inner workings of LuaCsForBarotrauma, so I'm not sure I can answer this. But in the vanilla game, simply enabling a mod will load all the necessary XML content. I'm also not sure which removed method you are referring to.
